### PR TITLE
RFC: React JSX component declaration syntax

### PR DESCRIPTION
**Pull request by thr0w**
```html
<!DOCTYPE JSX>

<ReactClass name="Timer">
  <div>Seconds Elapsed: {secondsElapsed}</div>

  <script>
    function getInitialState() {
      return { secondsElapsed = 0 };
    }
    function tick(){
      this.setState({secondsElapsed: secondsElapsed + 1});
    }
    function componentDidMount() {
      this.interval = setInterval(this.tick, 1000);
    }
    function componentWillUnmount() {
      clearInterval(this.interval);
    }
  </script>
</ReactClass>

React.render(<Timer />, mountNode);
```

=== 
Why
* Better look and feel (I think so)
* Better compatibility with IDE's, linters...
* 100% compatible with React legacy code

=== 
Done
* Enabling tag `<!DOCTYPE JSX>`. I don't like it, and maybe it's not necessary because JSXModule just works when JSX is as statement (not expression), but i'm worry with legacy code.
* Parser for script, link and style tags.
* visitors for jstransform
* visitors for react
* the code above works well

=== 
Todo
* avoid this from all know (class internals) identifiers
> how detect setState when invoking methods? (like push, shift, splice...)
* Transform css script into javascript objects.
> Rename css classes to allow local scope
> autoprex vendor tags
** I18N
** Docs

===
Repositories:
https://github.com/thr0w/esprima
https://github.com/thr0w/jstransform
https://github.com/thr0w/react

===
References:
* https://muut.com/riotjs/
* https://github.com/undoZen/htmlxify

### CODE REVIEW

**Comment by RusAlex** [comment]
Very simlar to WebComponents.

Huge job has been done.

**Comment by jimfb** [comment]
Personally, I like it; somehow it does feel cleaner and more declarative.  Especially when the common case can be compiled to a stateless function.

Having said that, I think the biggest "problem" is that we already support the ES6 class syntax, and the current plan is to also support the stateless function syntax.  These were chosen primarily because they are javascript standard constructs.  Having other officially supported syntactic alternatives makes the documentation fragmented, and since this is semantically equivalent, it's not clear to me that this improved syntax justifies the additional cognitive load for users.  For this reason, I suspect we are unlikely to take this change, even though I like the spirit of the change.

You had mentioned that this "works when JSX is as statement (not expression)".  Can you elaborate a little more?  When paired with something like https://www.npmjs.com/package/jsx-control-statements, does that mostly solve your concerns on this front?  Or is there a more subtle concern?

My intuition is that your best bet would be to package this up as a separate NPM package as a transpiler plugin (ideally a babel transform) that people can pull in (just like jsx-control-statements).  Then, people can vote for your syntax by using it in their own projects.  If it becomes sufficiently popular/useful, we can consider pulling it into React core.

Anyway, that's my two cents.  Doesn't necessarily reflect the wider opinions of the rest of the team.

**Comment by jimfb** [comment]
As a side note: We can only accept pull requests against master (this PR is against stable).  Normally we'd need to close it out (since it's not possible to change the merge branch of a PR), but I'm going to leave this one open since it's just an RFC anyway.

**Comment by yaycmyk** [comment]
Would be weird if it looked like this?

```html
<ReactClass name="Timer">
    <getInitialState>
        return { secondsElapsed = 0 };
    </getInitialState>

    <componentDidMount>
        this.interval = setInterval(this.tick, 1000);
    </componentDidMount>

    <componentWillUnmount>
        clearInterval(this.interval);
    </componentWillUnmount>

    <render>
        <div>Seconds Elapsed: {secondsElapsed}</div>
    </render>

    <tick>
        this.setState({secondsElapsed: secondsElapsed + 1});
    </tick>
</ReactClass>
```

maybe I'm missing the point of the OP. It's just having an arbitrary script block felt odd to me

**Comment by zpao** [comment]
We're not going to do this in any official way. Our goal is to make component definitions be JS as much as possible (with a little bit of sugar with JSX).

If you would like to maintain and support a custom transform tool, you are more than welcome to.

**Comment by thr0w** [comment]
@jimfb / @zpao , I asked for pull request because I wanna make you to know, and others, that I able to help make that change. I need help to choose best syntax, I need you guide me about react dev rules. I am aware that PR is unfinished.

Sorry about PR, I promise not do again. Nonetheless, I wanna know how to continue this conversation (issue?, another repository, group, slack) and, of course, if you wanna continue with this.

In other words, I have no expertise to make a good tool alone if you are available to guide me I believe I can make a good tool and then you might want a PR. How I need be guided? Use acorn / babel instead esprima (thanks @sebmck  ); use or not if / repeat directive like riotjs; someone is working in it ...

jim, when I said 'JSX is as statement' I mean to say JSX as AST ExpressionStatement.

sorry my english, I'm still learning.

**Comment by milesj** [comment]
This looks oddly similar to Riot. https://muut.com/riotjs/

**Comment by jimfb** [comment]
@thr0w Not a problem.  PRs are fine, they generally get more attention than issues, so keep up the good work.

As @zpao said, to the extent possible, we try to stick with standard javascript constructs, so this feature is not likely to make it into the core any time soon.

If you want to continue building out this type of feature, I'd encourage you to create a repository/npmModule just like https://github.com/valtech-au/jsx-control-statements and release it independently.

If you are looking to help contribute to React core, I'd recommend taking a look at one of these issues: https://github.com/facebook/react/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+bug%22
